# STOCKS ANALYSIS - A Data-Driven Solution

# Analysis of Stock Data

## Overview
This Jupyter Notebook analyzes the stock data from Mega Wholesalers LTD. The data is loaded from an Excel file, and various transformations are applied to understand stock movements.

## Data Source
The stock data is sourced from the file:
- File Path: C:\\Users\\jeann\\Downloads\\Mega_Wholesalers_LTD\\Sample_Data - S00062.xlsx

## Data Columns
The data columns are renamed for clarity:
- "Document Date" is renamed to "Date"
- "Item Code" is renamed to "SKU"
- "Total Qty IN" is renamed to "StockIn"
- "Total Qty OUT" is renamed to "StockOut"

## Data Processing
The DataFrame is sorted by "SKU" and "Date" to facilitate analysis. The "Date" column is converted to a datetime format.

A new column "StockOnHand" is created, representing the cumulative stock on hand, calculated as the cumulative sum of "StockIn" minus the cumulative sum of "StockOut".

## Libraries Used
- Pandas: Data manipulation library for handling tabular data.
- Matplotlib (if used): Library for creating visualizations.

## Notes
- Ensure that the file path is correct to load the data successfully.
- Verify the accuracy of the data transformations.

**Author:** Jean Claude Nshimiyimana
**Date:** 09/27/2023


In [1]:
import pandas as pd
file_path = "C:\\Users\\jeann\\Downloads\\Mega_Wholesalers_LTD\\Sample_Data - S00062.xlsx"

# Load the Excel file into a Pandas DataFrame
df = pd.read_excel(file_path)

# Rename columns 
df = df.rename(columns={
    "Document Date": "Date",
    "Item Code": "SKU",
    "Total Qty IN": "StockIn",
    "Total Qty OUT": "StockOut"
})

# Sort the DataFrame by SKU and Date
df["Date"] = pd.to_datetime(df["Date"])
df = df.sort_values(by=["SKU", "Date"])
df["StockOnHand"] = df["StockIn"].cumsum() - df["StockOut"].cumsum()

## Calculate aggregate statistics

In [2]:
df

Supplier Code       Date  Day  Month  Year  Branch     SKU  \
67573         S00062 2020-12-28   28     12  2020  ELD001  DET001   
212260        S00062 2020-12-28   28     12  2020  ELD002  DET001   
256198        S00062 2020-12-28   28     12  2020  NRB006  DET001   
76461         S00062 2020-12-29   29     12  2020  DAN001  DET001   
109275        S00062 2020-12-29   29     12  2020  DAN002  DET001   
...              ...        ...  ...    ...   ...     ...     ...   
210835        S00062 2023-09-25   25      9  2023  DAN001  WHI039   
259005        S00062 2023-09-25   25      9  2023  DAN002  WHI039   
11615         S00062 2023-09-26   26      9  2023  DAN001  WHI039   
21841         S00062 2023-09-26   26      9  2023  NRB001  WHI039   
144243        S00062 2023-09-27   27      9  2023  DAN002  WHI039   

                                                Item Name  StockIn  StockOut  \
67573   DETREX SOAP BLACK MEDICATED (PINE DROPS) 100GM...        0         1   
212260  DETREX SOAP BLACK MEDICATED (PINE DROPS) 100GM...        0         1   
256198  DETREX SOAP BLACK MEDICATED (PINE DROPS) 100GM...        0         2   
76461   DETREX SOAP BLACK MEDICATED (PINE DROPS) 100GM...        0         1   
109275  DETREX SOAP BLACK MEDICATED (PINE DROPS) 100GM...        0         2   
...                                                   ...      ...       ...   
210835              WHITE WASH PINK-SENSATION 175GM*48PCS        0         1   
259005              WHITE WASH PINK-SENSATION 175GM*48PCS        0         2   
11615               WHITE WASH PINK-SENSATION 175GM*48PCS        0         1   
21841               WHITE WASH PINK-SENSATION 175GM*48PCS        0         1   
144243              WHITE WASH PINK-SENSATION 175GM*48PCS        0         1   

        StockOnHand  
67573            -1  
212260           -2  
256198           -4  
76461            -5  
109275           -7  
...             ...  
210835       -86538  
259005       -86540  
11615        -86541  
21841        -86542  
144243       -86543  

[292341 rows x 11 columns]

In [3]:
df.describe()

Date            Day          Month  \
count                         292341  292341.000000  292341.000000   
mean   2022-05-05 03:07:20.473282816      15.585696       6.337883   
min              2020-12-28 00:00:00       1.000000       1.000000   
25%              2021-09-16 00:00:00       8.000000       4.000000   
50%              2022-04-30 00:00:00      15.000000       6.000000   
75%              2022-12-10 00:00:00      23.000000       9.000000   
max              2023-09-27 00:00:00      31.000000      12.000000   
std                              NaN       8.847420       3.341435   

                Year        StockIn       StockOut    StockOnHand  
count  292341.000000  292341.000000  292341.000000  292341.000000  
mean     2021.857567      19.428599      19.724633  -40005.613838  
min      2020.000000   -1500.000000   -1500.000000  -90711.000000  
25%      2021.000000       0.000000       1.000000  -58490.000000  
50%      2022.000000       0.000000       2.000000  -40836.000000  
75%      2022.000000       0.000000       6.000000  -15584.000000  
max      2023.000000   10815.000000    6825.000000    1033.000000  
std         0.765982     191.586093     108.207134   23941.904500

In [4]:

statistics = df.describe()

# Print the aggregate statistics
print("Aggregate Statistics:")
print(statistics)


Aggregate Statistics:
                                Date            Day          Month  \
count                         292341  292341.000000  292341.000000   
mean   2022-05-05 03:07:20.473282816      15.585696       6.337883   
min              2020-12-28 00:00:00       1.000000       1.000000   
25%              2021-09-16 00:00:00       8.000000       4.000000   
50%              2022-04-30 00:00:00      15.000000       6.000000   
75%              2022-12-10 00:00:00      23.000000       9.000000   
max              2023-09-27 00:00:00      31.000000      12.000000   
std                              NaN       8.847420       3.341435   

                Year        StockIn       StockOut    StockOnHand  
count  292341.000000  292341.000000  292341.000000  292341.000000  
mean     2021.857567      19.428599      19.724633  -40005.613838  
min      2020.000000   -1500.000000   -1500.000000  -90711.000000  
25%      2021.000000       0.000000       1.000000  -58490.000000  
50%    

# Calculate stock movements

# Cumulative Stock Analysis

## Overview
This section of the notebook calculates the cumulative stock in and stock out based on the provided stock data. The date column is converted to a datetime format for accurate analysis.


In [5]:

date_column = "Date"  
stock_in_column = "StockIn"  
stock_out_column = "StockOut" 
# Convert the date column to datetime format
df[date_column] = pd.to_datetime(df[date_column])
df["Cumulative Stock In"] = df[stock_in_column].cumsum()
df["Cumulative Stock Out"] = df[stock_out_column].cumsum()


# Identify Out-of-Stock Periods

## Overview
This section of the notebook identifies and calculates the duration of out-of-stock periods based on cumulative stock in and stock out data.


In [6]:
out_of_stock_periods = []
current_period = None
total_duration = 0  # Initialize total duration

for index, row in df.iterrows():
    if row["Cumulative Stock In"] < row["Cumulative Stock Out"]:
        if current_period is None:
            current_period = {"start_date": row[date_column]}
    elif current_period is not None:
        current_period["end_date"] = row[date_column]
        out_of_stock_periods.append(current_period)
        total_duration += (row[date_column] - current_period["start_date"]).days  # Calculate duration
        current_period = None


## Print Out-of-Stock Periods and Total Duration

In [7]:
# Print out-of-stock periods
for period in out_of_stock_periods:
    print(f"Out of stock from {period['start_date']} to {period['end_date']}")

# Print total duration of out-of-stock periods
print(f"Total duration of out-of-stock periods: {total_duration} days")

Out of stock from 2020-12-28 00:00:00 to 2020-12-30 00:00:00
Out of stock from 2021-02-08 00:00:00 to 2021-03-21 00:00:00
Out of stock from 2021-04-16 00:00:00 to 2021-07-11 00:00:00
Out of stock from 2021-07-13 00:00:00 to 2021-08-07 00:00:00
Out of stock from 2021-08-10 00:00:00 to 2021-08-14 00:00:00
Out of stock from 2022-02-09 00:00:00 to 2022-04-03 00:00:00
Total duration of out-of-stock periods: 211 days


# Inventory Analysis Report

## Introduction

I am pleased to present the results of our recent inventory analysis, conducted to address the challenges faced by FMCG companies in stock control, as highlighted in our previous discussions.

## Key Findings

- **Out-of-Stock Periods:** Our analysis identified several instances of out-of-stock periods over the analyzed period of three years. Notably, the total duration of these stockouts amounted to 211 days.

- **Operational Impact:** These out-of-stock periods had a significant impact on our operations, affecting our ability to meet customer demand and potentially resulting in lost sales opportunities.

## Implications and Recommendations

The insights gleaned from this analysis are invaluable for optimizing our inventory management and ensuring product availability. By addressing the identified challenges, we can enhance operational efficiency, financial performance, and customer satisfaction.

### Action Steps

1. **Inventory Optimization:** We should consider adjusting reorder points, safety stock levels, and supply chain practices to minimize stockouts and enhance product availability.

2. **Customer Satisfaction:** Improving our stock management practices will directly benefit customer satisfaction, fostering trust and loyalty among our clientele.

3. **Financial Performance:** Reducing stockout instances will positively impact our financial performance by capturing more sales and reducing revenue losses.

4. **Operational Efficiency:** Streamlining inventory operations through enhanced stock control will lead to increased operational efficiency.

## Conclusion

In conclusion, this analysis offers actionable insights to enhance our inventory management strategies. Addressing the identified challenges will not only improve our operational efficiency but also contribute to our success and sustainability in the highly competitive FMCG market.


# Inventory Analysis API

## Overview
This Flask API analyzes inventory data, identifies out-of-stock periods, and provides relevant insights. It exposes a single endpoint `/analyze` for data analysis based on user-defined parameters.

## Data Source
The API loads inventory data from an Excel file, processes missing values, and calculates cumulative stock in and stock out.

## API Endpoint
- **Endpoint:** `/analyze`
- **Method:** POST
- **Request Parameters:**
  - `startDate` (str): Start date for analysis in "dd/mm/yyyy" format.
  - `endDate` (str): End date for analysis in "dd/mm/yyyy" format.
  - `branch` (str, optional): Branch filter.
  - `itemCode` (str, optional): Item code filter.

## Data Analysis
- Out-of-stock periods are identified based on cumulative stock in and stock out data.
- Results include total purchases, total sales, actual out-of-stock days, and average out-of-stock days.

## Running the API Locally
1. **Requirements:**
    - Python
    - Flask (`pip install Flask`)

2. **Run the API:**
   - Navigate to the project directory containing the API code.
   - Execute the following command in the terminal:
     ```bash
     python your_api_file.py
     ```
     Replace `your_api_file.py` with the actual filename.

3. **Local Server Output:**
   - Upon successful execution, you'll see output similar to:
     ```
     * Serving Flask app '__main__'
     * Debug mode: off
     WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
     * Running on http://127.0.0.1:5000
     Press CTRL+C to quit
     ```

## Exposing to the Public Internet with ngrok
1. **Install ngrok:**
    - Follow the instructions on [ngrok's website](https://ngrok.com/download) to install ngrok.

2. **Expose Locally Running API:**
    - In a new terminal window, navigate to the directory where ngrok is installed.
    - Run the following command to expose the local Flask app:
      ```bash
      ngrok http 5000
      ```
      (Assuming your Flask app is running on the default port 5000)

3. **Accessing the Public API:**
   - Ngrok generates a public URL (e.g., `http://abc123.ngrok.io`). Use this URL to interact with your API from anywhere.


In [8]:
#pip install Flask

In [9]:
import pandas as pd
from flask import Flask, request, jsonify

app = Flask(__name__)

# Load the Excel file into a Pandas DataFrame
file_path = "C:\\Users\\jeann\\Downloads\\Mega_Wholesalers_LTD\\Sample_Data - S00062.xlsx"
df = pd.read_excel(file_path)

# Rename columns 
df = df.rename(columns={
    "Document Date": "Date",
    "Item Code": "SKU",
    "Total Qty IN": "StockIn",
    "Total Qty OUT": "StockOut"
})

# Handle missing data 
df = df.dropna()

# Sort the DataFrame by SKU and Date
df["Date"] = pd.to_datetime(df["Date"])
df = df.sort_values(by=["SKU", "Date"])
df["StockOnHand"] = df.groupby(["SKU"])["StockIn"].cumsum() - df.groupby(["SKU"])["StockOut"].cumsum()

# Define constants for column names
date_column = "Date"
stock_in_column = "StockIn"
stock_out_column = "StockOut"

# Convert the date column to datetime format
df[date_column] = pd.to_datetime(df[date_column])
df["Cumulative Stock In"] = df.groupby(["SKU"])[stock_in_column].cumsum()
df["Cumulative Stock Out"] = df.groupby(["SKU"])[stock_out_column].cumsum()

@app.route('/analyze', methods=['POST'])
def analyze_stock():
    try:
        # Extract data from the POST request
        data = request.get_json()

        # Extract parameters from the request
        start_date = pd.to_datetime(data["startDate"], format="%d/%m/%Y")
        end_date = pd.to_datetime(data["endDate"], format="%d/%m/%Y")
        branch = data.get("branch")
        item_code = data.get("itemCode")

        # Data validation
        if start_date > end_date:
            return jsonify({"error": "Start date cannot be after end date."}), 400

        # Apply additional filters
        if branch:
            filtered_df = df[df["Branch"] == branch]
        else:
            filtered_df = df.copy()

        if item_code:
            filtered_df = filtered_df[filtered_df["SKU"] == item_code]

        filtered_df = filtered_df[(filtered_df[date_column].dt.year == 2023)]

        out_of_stock_periods = []
        current_period = None
        total_duration = pd.Timedelta(0)

        for sku, sku_data in filtered_df.groupby("SKU"):
            stock_in_cumulative = sku_data["Cumulative Stock In"]
            stock_out_cumulative = sku_data["Cumulative Stock Out"]
            date_values = sku_data[date_column]

            for i in range(len(date_values)):
                if stock_in_cumulative.iloc[i] < stock_out_cumulative.iloc[i]:
                    if current_period is None:
                        current_period = {"start_date": date_values.iloc[i]}
                elif current_period is not None:
                    current_period["end_date"] = date_values.iloc[i]
                    out_of_stock_periods.append(current_period)
                    total_duration += (current_period["end_date"] - current_period["start_date"])
                    current_period = None

        # Calculate average out of stock days
        num_periods = len(out_of_stock_periods)
        average_oosd = total_duration.days / num_periods if num_periods > 0 else 0

        # Prepare the results
        results = {
            "itemCode": item_code,
            "branch": branch,
            "itemData": {
                "startDate": start_date.strftime("%d/%m/%Y"),
                "endDate": end_date.strftime("%d/%m/%Y"),
                "totalPurchases": int(filtered_df[stock_in_column].sum()),  # Convert to int
                "totalSales": int(filtered_df[stock_out_column].sum()),  # Convert to int
                "actualOOSD": total_duration.days,
                "averageOOSD": round(average_oosd, 2)
            }
        }

        # Return the results as JSON
        return jsonify(results), 200

    except Exception as e:
        return jsonify({"error": str(e)}), 500

In [ ]:
if __name__ == '__main__':
    app.run()


 * Serving Flask app '__main__'
 * Debug mode: off


 * Running on http://127.0.0.1:5000
Press CTRL+C to quit
